[MISC] Ultrasound simulation model

Rakshith2597 opened the pull request:
Submitting training module for breast ultrasound simulation model.

This is part of the project MIRIAD: Many Incarnations of Screening of Radiology for High Throughput Disease Screening via Multiple Instance Reinforcement Learning with Adversarial Deep Neural Networks, sponsored by INTEL TECHNOLOGY INDIA PVT. LTD.

Principal Investigators:

Dr Debdoot Sheet (PI), Dr Nirmalya Ghosh (Co-PI)
Department of Electrical Engineering
Indian Institute of Technology Kharagpur

Dr Ramanathan Sethuraman (Co-PI)
Intel Technology India Pvt. Ltd.

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/README.md:
```diff
@@ -0,0 +1,101 @@
+# Multi-frequncy Ultrasound Simulation
+
+Ultrasound simulators are used by clinicians and by the scientific community of practitioners and educators. Ultrasound image formation depends on factors like frequency and size of the transducer, tissue characteristics, nonlinear attenuation, diffraction, and scattering due to the medium. Prior approaches try to incorporate various factors by modelling their physics and solving wave equations,  ...
+We further derive the relationship between the frequency of the transducer and the resolution of the ultrasound image to be able to simulate and form images at frequencies other than at which the model is trained. This is achieved by interpolation between dilated kernels in a multi-linear separable CNN used to control the resolution of the generated image allowing for zero-shot adaptation. Using  ...
```
```suggestion
We further derive the relationship between the frequency of the transducer and the resolution of the ultrasound image to be able to simulate and form images at frequencies other than at which the model is trained. This is achieved by interpolation between dilated kernels in a multi-linear separable CNN used to control the resolution of the generated image allowing for zero-shot adaptation. Using the multi-linear separable design, we can simulate entire 3D volumes by training the model only on 2D ultrasound images. 
```

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/dataloader.py:
```diff
@@ -0,0 +1,165 @@
+import torch
+
+
+device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
```
unused variable

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/solve.py:
```diff
@@ -0,0 +1,358 @@
+import time
+from model import GeneratorModel, DiscriminatorModel, GeneratorInter, Generator3dInter
+
+device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
```
It would be better to pass device as one of the arguments rather then hardcode it as a global variable.

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/solve.py:
```diff
@@ -0,0 +1,358 @@
+                      end=" ")
+
+                # tensorboard logging
+                if(self.log):
```
```suggestion
                if self.log:
```

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/solve.py:
```diff
@@ -0,0 +1,358 @@
+                        self.writer.add_scalar('LR', get_lr(
+                            self.g_opt), i_batch + len(self.train_data) * epoch)
+
+                    if(i_batch == 0 and epoch % self.args.vis_step == 0):
```
```suggestion
                    if i_batch == 0 and epoch % self.args.vis_step == 0 :
```
No need to add parenthesis in if condition

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/solve.py:
```diff
@@ -0,0 +1,358 @@
+
+
+class solver_inter2d(solver):
+    def __init__(self, args, train_data=None, test_data=None, restore=1):
```
You inherit `solver` class but do not call its `__init__` method:
`super().__init__(args, train_data=None, test_data=None, restore=1)`
Adding this call would allow not to duplicate common code. 
You will need to extract common part of all successor classes (`solver_inter2d` and `solver_inter3d`) into `solver`'s `__init__` method and add to the successors' `__init__`'s only differing functionality. This would make code more readable, compact and easier to maintain. [Learn more](https://docs.python.org/3/library/functions.html#super) about python's super.

morkovka1337 reviewed (changes requested):
Nice work! It would be even better if you add:
* Some demo application which would allow other users to see what this project does
* Export code for exporting Pytorch models to the ONNX and OpenVINO IR formats. You provide exported models, but it is unclear, how to get them. If someone forks this project and changes\adds some functionality, it could be difficult to figure out how to get exported models.
* Minimal tests which cover:
  * Training code
  * Testing code - test if provided model reproduces some reference metric
  * Export code - test if the model could be exported to ONNX & OpenvinoIR and reference metric is reproduced (maybe with small divirgence)

By the way, could you provide some information about what metric is used to evaluate quality of the model and give reference values?

brmarkus commented on misc/pytorch_toolkit/breast_ultrasound_simulation/README.md:
````diff
@@ -0,0 +1,101 @@
+                        .
+                        .
+```
+Segmentation masks from [INbreast](http://medicalresearch.inescporto.pt/breastresearch/index.php/Get_INbreast_Database) dataset is used to generate stage 0 simulation results for breast ultrasound.
````
The link to "http://medicalresearch.inescporto.pt/breastresearch/index.php/Get_INbreast_Database" seems broken.

Rakshith2597 replied:
We have provided an alternate source for the dataset in the readme file. Note that, we have not used the INbreast dataset to train the model. We demonstrate the ability to simulate breast ultrasound images using a model trained on the IVUS dataset.

brmarkus reviewed (commented):
Interested in this topic due to personal matter.

Rakshith2597 commented:
- Added a demo application that would allow other users to see what this project does.
- Export code for exporting Pytorch models to the ONNX and OpenVINO IR formats has been added. 
- Minimal unit tests added. 
- Details regarding metric used to evaluate the quality of model has been provided in ReadMe file.

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/README.md:
````diff
@@ -0,0 +1,137 @@
+                        .
+                        .
+```
+Segmentation masks from [INbreast](http://medicalresearch.inescporto.pt/breastresearch/index.php/Get_INbreast_Database) dataset is used to generate stage 0 simulation results for breast ultrasound. Alternate source for INbreast dataset [link](https://www.kaggle.com/ramanathansp20/inbreast-dataset). Note that, we have not used INbreast dataset to train these models. We demonstrate the ability to s ...
````
The first link (medical research) is dead. No need to store it, you can remove it.

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/metric.py:
```diff
@@ -0,0 +1,65 @@
+		for j in range(s):
+			n = np.linalg.norm(c - np.array([i, j]))
+			if(n <hf and n>lf):
+				mask[i][j] = 1
+
+	return mask
+
+def fft(img, mask, v):
+	img = img/255
+	m = np.mean(img)
+	img = (img - m)
+	f = np.fft.fft2(img)
+	f = np.fft.fftshift(f)
+	f = mask*f
+	s = np.sum(np.abs(f)**2)
+
+	return s/m,f
+
+
+DIR = [args.dir]
+
+# DIR.sort()
+print(DIR)
+
+
+sam_avg = [0]*len(DIR)
+fft_avg = [0]*len(DIR)
+s = 128
+lf = s/8
+hf = s/4
+mask = filter(s, lf, hf)
+for j, D in enumerate(tqdm(DIR)):
+	images = [f for f in os.listdir(DIR[j])]
+	for i, im in enumerate(images):
+		
+		a = cv2.imread(os.path.join(D, im),0)
+		fa = fft(a, mask, hf - lf)
+		sam_avg[j] += fa[0]
+		
+		fft_avg[j] = (fft_avg[j]*i + fa[1])/(i+1)
```
Please, use the same indentation size everywhere.

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/model.py:
```diff
@@ -0,0 +1,425 @@
+                0, 0, 0), stride=(
+                1, 1, 1), groups=ch, bias=True)
+
+        """if(wts_list is not None):
+            print("In asssss")
+            self.xdim3d.weight = torch.nn.Parameter(
+                torch.unsqueeze(wts_list.xdim.weight, 2))
+            self.zdim3d.weight = torch.nn.Parameter(
+                torch.unsqueeze(wts_list.xdim.weight, 2).permute(0,1,4,3,2))
+            self.bias3d.weight = torch.nn.Parameter(
+                torch.unsqueeze(wts_list.bias.weight, 2))
+            self.bias3d.bias = torch.nn.Parameter(
+                wts_list.bias.bias)"""
```
What is it?

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/solve.py:
```diff
@@ -0,0 +1,335 @@
+print(torch.cuda.is_available())
+
+
+class solver():
```
Class names should start with [capitals.](https://www.python.org/dev/peps/pep-0008/#class-names)

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/solve.py:
```diff
@@ -0,0 +1,335 @@
+            print("restored")
+
+    def set_requires_grad(self, net, requires_grad=False):
+        """Set :requies_grad=Fasle for all the networks to avoid unnecessary computations
```
```suggestion
        """Set :requires_grad=Fasle for all the networks to avoid unnecessary computations
```

By the way, it is strange to see docstring for only some functions. Add docstring to other functions and classes or remove them at all.

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/solve.py:
```diff
@@ -0,0 +1,335 @@
+                AdvD_epoch += self.adversarial_loss.item()
+                # print(time.time())
+                # Python Logging
+                print('\rStep [{}/{}], L1: {:.4f}, AccD: {:.4f}, AdvD: {:.4f}, AdvG: {:.4f} Time: {:.1f}sec'.format((i_batch + 1),
+                                                                                                                    len(self.train_data),
+                                                                                                                    L1_epoch / (i_batch + 1),
+                                                                                                                    DisAcc_epoch / (i_batch + 1),
+                                                                                                                    AdvD_epoch / (i_batch + 1),
+                                                                                                                    AdvG_epoch / (i_batch + 1),
+                                                                                                                    time.time() - start_time),
+                      end=" ")
```
Line is too long and could be displayed weird. Please, format this to have maximum line length 120 symbols.

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/solve.py:
```diff
@@ -0,0 +1,335 @@
+            print('Step:{}'.format(i_batch))
+
+
+class solver_inter2d(solver):
```
Same for class name

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/solve.py:
```diff
@@ -0,0 +1,335 @@
+            print("Loaded Model for inferencing IVUS2D")
+
+
+class solver_inter3d(solver):
```
class name

morkovka1337 commented on misc/pytorch_toolkit/rakshith@203.110.240.161/README.md:
```diff
@@ -0,0 +1,242 @@
+# Deep Neural Network for Chest X-Ray Screening
```
Strange file path (rakshith@203.110.240.161). Accidently added?)

Rakshith2597 replied:
Yes, It was an accident. Sorry for that.

morkovka1337 reviewed (changes requested):
Please, check files with IP address in their names.
Move your toy dataset to `data/` folder and in root of this repo.
As for empty folders with .keep files (logs, data, checkpoints), you can create these folders when you directly need it. See https://docs.python.org/3/library/os.html#os.makedirs

morkovka1337 commented:
Run tests

morkovka1337 commented:
Please, fix pylint issues. You can further check this locally: `pip install pylint` and `pylint --rcfile ../../../.pylintrc -v src && pylint --rcfile ../../../.pylintrc -v tests` when you are in your project directory (breast_ultrasound_simulation).
Also, there is no setup.py file that allows to install your training code as a python package via pip.

morkovka1337 commented:
Looks like you have accidently uploaded a lot of images instead of toy dataset...

Rakshith2597 commented:
> Looks like you have accidentally uploaded a lot of images instead of a toy dataset...

Actually, it was not a mistake. 
Our model can simulate full 3D volumes using its multi-linear separable design by training the model only on 2D ultrasound images. To be precise, Given a CNN trained on a 2D ultrasound dataset acquired using a transducer of a particular frequency, we can simulate 2D ultrasound images and 3D volumes corresponding to various transducer frequencies without the need for availability of additional training data in 3D and at a different frequency. 
In the dummy/toy dataset provided, the folder _'real_images'_ contains 2D images and '_real_images_3d_' contains images corresponding to a **single** 3D volume. If **a user wishes to** evaluate the performance of 3D simulation, they can use the sample/toy 3D volume provided. That was the reason for keeping those images (folders: real_images_3d and stage0_3d).  Note that these images are not mandatory for the unit tests since the model is trained only on 2D images. All evaluations provided in the code( loss, performance metric, etc.) are done on 2D images. If needed, I can provide the files belonging to the 3D volume in google drive and mention it in the ReadMe.

morkovka1337 commented:
> If needed, I can provide the files belonging to the 3D volume in google drive and mention it in the ReadMe.

I think, this would be better option. We store only minimal example of data in git, and data more than, like 10-20 images, is more preferable to store on the GDrive.

Ilya-Krylov commented:
@Rakshith2597 will you update it?

Rakshith2597 commented:
> will you update it?

@Ilya-Krylov 
Yes, I will be. The proposed model was trained using the IVUS dataset which does not comply with the licensing criteria of training_extension. To abide by the required license clauses, we have retrained the model with a new dataset. At the moment we are working on reorganising the code and adding necessary unit tests. We plan to update this as soon as necessary internal clearances are obtained.

Rakshith2597 commented:
@morkovka1337 The master and develop branch has changed significantly from the previous version. I see that this PR has been rebased to the `misc` branch. 
1. Shall I proceed with this PR to merge with the misc branch?
2. There are a few more models which are part of the project that we had planned to submit here. Shall I create a PR on the `misc` branch for those as well?

morkovka1337 commented:
> @morkovka1337 The master and develop branch has changed significantly from the previous version. I see that this PR has been rebased to the `misc` branch.
> 
> 1. Shall I proceed with this PR to merge with the misc branch?

Yes, non-templated models are in the separate `misc` branch now.
> 2. There are a few more models which are part of the project that we had planned to submit here. Shall I create a PR on the `misc` branch for those as well?

Yes

Rakshith2597 commented:
@morkovka1337 Was this reviewed?

morkovka1337 commented:
Sorry, did not see ready for review mark. Will review in the nearest future.

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/README.md:
```diff
@@ -0,0 +1,141 @@
+
+Ultrasound simulators are used by clinicians and by the scientific community of practitioners and educators. Ultrasound image formation depends on factors like frequency and size of the transducer, tissue characteristics, nonlinear attenuation, diffraction, and scattering due to the medium. Prior approaches try to incorporate various factors by modeling their physics and solving wave equations wh ...
+We further derive the relationship between frequency of the transducer and resolution of the ultrasound image in order to be able to simulate and form images at frequencies other than at which the model is trained. This is achieved by interpolation between dilated kernels in a multi-linear separable CNN that is used to control the resolution of the generated image allowing for zero-shot adaptatio ...
+Given a  CNN trained on a  2D ultrasound dataset acquired using a transducer of a particular frequency, we can simulate  ultrasound images corresponding to various transducer frequencies without the need for availability of additional training data at different frequencies. We have performed an  extensive validation of our approach using both real and simulated ultrasound images. Using a regressi ...
```
```suggestion
Ultrasound simulators are used by clinicians and by the scientific community of practitioners and educators. Ultrasound image formation depends on factors like frequency and size of the transducer, tissue characteristics, nonlinear attenuation, diffraction, and scattering due to the medium. Prior approaches try to incorporate various factors by modeling their physics and solving wave equations which leads to computationally expensive solutions. In this work, we propose a fast simulation approach using Convolutional Neural Network (CNN) to model the non-linearity of signal interaction with ultrasound image formation. The network is trained in an adversarial manner using a Turing test discriminator. The simulation process consists of two stages, Stage 0 is a pseudo B-mode simulator that provides an initial estimate of the speckle map, Stage 1 is a multi-linear separable CNN which further refines the speckle map to provide the final output. In order to ensure reproducibility, we have initialized the experiments with predefined random seeds.
We further derive the relationship between frequency of the transducer and resolution of the ultrasound image in order to be able to simulate and form images at frequencies other than at which the model is trained. This is achieved by interpolation between dilated kernels in a multi-linear separable CNN that is used to control the resolution of the generated image allowing for zero-shot adaptation.
Given a CNN trained on a 2D ultrasound dataset acquired using a transducer of a particular frequency, we can simulate  ultrasound images corresponding to various transducer frequencies without the need for availability of additional training data at different frequencies. We have performed an  extensive validation of our approach using both real and simulated ultrasound images. Using a regression analysis we have concluded that there exists a linear relationship between the proposed kernel interpolation factor and the transducer frequency, thereby laying the foundation of a multi-frequency ultrasound simulator.
```

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/README.md:
```diff
@@ -0,0 +1,141 @@
+### Prerequisites
+ * python 3.7.3
+ * torch                  1.10.0+cu102
+ * torchvision            0.6.0+cu101
```
Aren't the requirements a bit confusing (different CUDA versions)?

morkovka1337 replied:
BTW, since you have requirements.txt, I think, this section can be deleted. I doubt that only the specific version of python is needed for this code to work (at least, it should work for the recent python 3 versions like 3.6+).

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/init_venv.sh:
```diff
@@ -0,0 +1,35 @@
+
+path_openvino_vars="${INTEL_OPENVINO_DIR:-/opt/intel/openvino}/bin/setupvars.sh"
+if [[ -e "${path_openvino_vars}" ]]; then
+  echo ". ${path_openvino_vars}" >>${venv_dir}/bin/activate
```
If you use OpenVINO as python package from PyPi, this is no more needed.

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/setup.py:
```diff
@@ -0,0 +1,4 @@
+from setuptools import setup, find_packages
+
+setup(name='breast_ultrasound_simulation',
+      packages=find_packages())
```
```suggestion
      packages=find_packages())
 
```
add newline

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/dataloader.py:
```diff
@@ -0,0 +1,82 @@
+
+        # images are the output, y
+        y = Image.open(os.path.join(self.Images_DIR, file_name))
+        # if self.resize:
+        #     y = y.resize((128, 128))
```
If you do not need this code, you can delete it.

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/dataloader.py:
```diff
@@ -0,0 +1,82 @@
+        return X, y, file_name
+
+
+class BUS_dataset(data.Dataset):
```
The class is very similar to the above one. You can create a base class and move common functionality like `__len__` (and `__init__` since they are almost the same if we do not pay attention to the field names) to it.

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/exporter.py:
```diff
@@ -0,0 +1,45 @@
+
+def export_model_onnx(config_args):
+    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    print(torch.cuda.is_available())
```
Either delete this or make it more informative (like `print(f'torch CUDA is available: {torch.cuda.is_available()}')`)

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/exporter.py:
```diff
@@ -0,0 +1,45 @@
+    state_dict = torch.load(model_path)
+    gen_old.load_state_dict(state_dict["generator1_weights"])
+    onnx_model = config_args['onnx_model_name']
+    dummy_input = torch.randn(1,1,142,142).cuda()
```
In the dataset class you resize images to 128x128. Is there a typo or some more complicated logic behind that?

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/exporter.py:
```diff
@@ -0,0 +1,45 @@
+    if config_args.onnx:
+        export_model_onnx(config_args)
+    if config_args.ir:
+        export_model_ir(config_args)
```
I think, you should check that onnx model exists before converting it to the IR. If it does not, `export_model_onnx` should be called.

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/infer.py:
```diff
@@ -0,0 +1,36 @@
+from utils.read_config import load_config
+
+
+net_config_path = "configs/net_config.json"
```
Please, transform this to a callable script (with `if __name__ == '__main__':` and arguments parsing)

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/model.py:
```diff
@@ -0,0 +1,419 @@
+        # do y x first follwed by the other spectal decompsed filters!
+
+        op = self.ydim(x)
+        #print(op.shape, x.shape)
```
commented code should be removed

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/model.py:
```diff
@@ -0,0 +1,419 @@
+        index = torch.Tensor([i for i in range(0, self.kernel, dil)]).to(torch.int64)
+        z = torch.zeros(ch, ch, self.kernel, 1)
+        if wts_list is None:
+            print("Not implemented")
```
If the code should not be executed further, then `NotImplementedError` should be raised. You later access a field of this object, so, I guess, `assert wts_list is not None, "some informative message if it is"` would be more applicable here.

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/model.py:
```diff
@@ -0,0 +1,419 @@
+                1, 1, 1), groups=ch, bias=True)
+
+        """if(wts_list is not None):
+            print("In asssss")
```
What should this mean?

Rakshith2597 replied:
Sorry. This was to be deleted before PR. Removed it.

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/model.py:
```diff
@@ -0,0 +1,419 @@
+
+    def forward(self, x):
+
+        # F.conv3d(x, self.wts_norm, padding=(0, self.pad, 0), stride=1)
```
same as above about commented code

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/model.py:
```diff
@@ -0,0 +1,419 @@
+
+        opx = self.xdim(op)
+        opz = self.zdim(op)
+        #print(select_x, op.shape)
+        # print(select_z)
```
same as above about commented code

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/model.py:
```diff
@@ -0,0 +1,419 @@
+            nn.Sigmoid()
+        )
+
+    def forward(self, x):
+        x = self.down1(x)
+        x = self.res1(x)
+        x = self.res2(x)
+        x = self.res3(x)
+        x = self.res4(x)
+        x = self.down2(x)
+
+        return x
```
Try to move common parts of the Generator model classes to one super-class. For example, `forward` method is the same and, I guess, some parts of the `__init__` can be unified also.

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/src/solve.py:
```diff
@@ -0,0 +1,387 @@
+
+    return model
+
+class solver_inter2d():
```
same as above about class names

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/tests/test_1_train.py:
```diff
@@ -0,0 +1,41 @@
+import os
+# import sys
+from torch.utils import data
+# sys.path.append("../src")
```
same as above about commented code

morkovka1337 commented on misc/pytorch_toolkit/breast_ultrasound_simulation/tests/test_1_train.py:
```diff
@@ -0,0 +1,41 @@
+import unittest
+import os
+# import sys
```
same as above about commented code

morkovka1337 reviewed (changes requested):
Somehow some previous comments were marked as resolved, though nothing was changed there. Please, fix them also.

nervana-ff commented:
Can one of the admins verify this patch?

ryanloney commented:
@goodsong81  can your team look into this?

Rakshith2597 commented:
@goodsong81 
Apologies for the delay in closing this PR. Resolved issues and incorporated the suggestions from previous maintainers as well. I have fixed issues raised in codacy as well (all issues expect the ones raised for readme file.).

Rakshith2597 commented:
Test results on local server.
![Screenshot from 2023-03-06 14-47-59](https://user-images.githubusercontent.com/25449712/223068177-a6688b1b-7cbf-42e0-b28e-b5a69335a274.png)

Pylint output
![Screenshot from 2023-03-06 15-02-38](https://user-images.githubusercontent.com/25449712/223074082-e5a6ed25-cf29-4be1-a310-c8f8d22a9bba.png)
